keps/sig-scheduling/5027-dra-admin-controlled-device-attributes/README.md

```diff
@@ -0,0 +1,848 @@
+it is preserved even when a ResourceSlice gets removed or updated. To achieve
+this, a new cluster-scoped ResourceSliceOverride type gets added. A single
+ResourceSliceOverride object specifies device attributes that apply to all
+devices matching a CEL expression, i.e. the same way as users select devices in
+a ResourceClaim.
+
+The scheduler must merge these additional attributes with the ones provided by
+the DRA drivers on-the-fly while it gathers information about available
+devices.
+
+### Notes/Constraints/Caveats
+
+Users who look at ResourceSlices to figure out which devices are available also
+need to consider ResourceSliceOverrides to get the full picture. Copying from
+the ResourceSliceOverride spec into the ResourceSlice status could help here,
+but would not be instantaneous and potentially cause write amplification (one
+ResourceSliceOverride affecting many different devices) and therefore is not
+part of this proposal.
+
+Perhaps `kubectl describe resourceslices` can be extended to include the
+additional information. For now this is out of scope.
+
+Creating a ResourceSliceOverride is racing with on-going scheduling attempts,
+which is unavoidable.
+
+### Risks and Mitigations
+
+From a security perspective, permission to override device attributes is
+expected to be limited to privileged users who already have the ability to add
+or remove DRA drivers, so there won't be a substantial difference. Normal users
+can only decide for themselves whether they ignore taints and do so at their
+own risk. They cannot break other users' workloads.
+
+Performance in the scheduler could be an issue. This will be mitigated by
+(re-)applying overrides only when they or the device definitions change, which
+should be rare.
+
+## Design Details
+
+### API
+
+The ResourceSliceOverride is a cluster-scoped type in the `resource.k8s.io` API
+group, initially in `v1alpha3` (the alpha version in Kubernetes 1.32). Because
+it may be useful to clean up after disabling the feature and because the
+device taint feature also uses this type, it gets served unconditionally as long as
+the `v1alpha3` version is enabled. Fields related specifically to this KEP
+are feature-gated.
+
+```Go
+type ResourceSliceOverride struct {
+metav1.TypeMeta
+    // Standard object metadata
+    // +optional
+    metav1.ObjectMeta
+
+    // Changing the spec automatically increments the metadata.generation number.
+    Spec ResourceSliceSpec
+}
+
+type ResourceSliceOverrideSpec struct {
+    // Devices defines how to override device attributes and taints.
+    Devices DeviceOverride
+}
+
+// DeviceOverride selects one or more devices by class, driver, pool, device names
+// and/or CEL selectors. All of these criteria must be satisfied by a device, otherwise
+// it is ignored by the override. A DeviceOverride with no selection criteria is
+// valid and matches all devices.
+type DeviceOverride struct {
```

Comment:
I wonder if the API might be more human-readable by moving the set of filters/selectors into a separate parent struct, so that the actual override data (`Attributes`/`Capacity`) is distinctly organized, and more easily disambiguated.

Reply:
I moved them into a `DeviceOverrideSelector` struct which is stored in a `Selector` field. The actual YAML then would look like this:

```YAML
apiVersion: resource.k8s.io/v1alpha3
kind: ResourceSliceOverride
metadata:
  ...
spec:
  devices:
    selector:
      driver: dra.example.com
      pool: work-node
      device: gpu-0
    attributes:
      my-additional-attribute-foo: bar
```

Reply:
The word "selector" for the field becomes a bit more problematic when considering CEL selectors inside it:

```yaml
apiVersion: resource.k8s.io/v1alpha3
kind: ResourceSliceOverride
metadata:
  ...
spec:
  devices:
    selector:
      deviceClass: dra.example.com
      selectors:
      - cel:
          expression: device.attributes["dra.example.com"].uid == "ABCD-1234"
    taints:
    - key: dra.example.com/unhealthy
      value: memory checksum error
      effect: NoSchedule
```

Selectors inside a selector? Hmm...

Reply:
How about

```
spec:
  devices:
    filters:
      deviceClass: dra.example.com
      selectors:
      - cel:
          expression: device.attributes["dra.example.com"].uid == "ABCD-1234"
```

Reply:
Works for me.

Reply:
Should it be "filters" or "filter"? It's not a list.

Reply:
Probably "filter" is the best compromise. A single "filter" can be a composition of a set of selectors, right? Even if we're playing loose and fast w/ grammar, I agree that "filters" suggests a list/array/slice instead of a dictionary.

Reply:
Now it's "filter" and `DevicePatchFilter`.